src/rules.jl

```diff
@@ -62,7 +62,10 @@ accumulate(Δ, rule::AbstractRule, args...) = add(Δ, rule(args...))
 """
 TODO
 """
-accumulate!(Δ, rule::AbstractRule, args...) = materialize!(Δ, broadcastable(add(cast(Δ), rule(args...))))
+function accumulate!(Δ, rule::AbstractRule, args...)
+    return materialize!(Δ, broadcastable(add(cast(Δ), rule(args...))))
+end
+accumulate!(Δ::Real, rule::AbstractRule, args...) = accumulate(Δ, rule, args...)
```

Comment:
Seems reasonable. Could probably expand this to `Δ::Number`, even. Could we [add a space between these two definitions](https://github.com/jrevels/YASGuide#linealignmentspacing-guidelines)? 🙂

I guess one place where this might yield unintuitive behavior is if it causes users to think that passing in a non-`materialize!`-able `Δ` always "just" works, e.g. if you call `accumulate!(::SArray, ...)` a user might expect it to fallback to `accumulate` instead of hitting the `setindex!` error it would currently hit when `materialize!(::SArray, ...)` is called. It's not immediately clear how to implement something that ensures that. However, I think we can justify/explain this fallback by saying that it only exists so that callers can handle things generically without checking for the scalar case. That way, it makes sense that numbers are special here, and avoids the impression that a similar fallback exists for immutable containers in general. 

Downstream packages are allowed to special-case these methods as well, which makes it even more okay for us to do so. I'm writing docstrings for these methods today which will hopefully make things clearer.

Reply:
> Could we add a space between these two definitions?

ofc :)

> Could probably expand this to Δ::Number

Sounds reasonable.

I'm on board with the rest of what you've suggested as well